test/evp_test.c

```diff
@@ -2354,14 +2353,73 @@ static int pkey_test_init(EVP_TEST *t, const char *name,
         EVP_PKEY_free(pkey);
         return 0;
     }
-    kdata->keyopinit = keyopinit;
-    kdata->keyop = keyop;
     if (!TEST_ptr(kdata->ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, propquery))) {
         EVP_PKEY_free(pkey);
         OPENSSL_free(kdata);
         return 0;
     }
     t->data = kdata;
+    return 1;
+}
+
+static int pkey_test_init(EVP_TEST *t, const char *name,
+                          int use_public,
+                          int (*keyopinit) (EVP_PKEY_CTX *ctx,
+                                            const OSSL_PARAM params[]),
+                          int (*keyop)(EVP_PKEY_CTX *ctx,
+                                       unsigned char *sig, size_t *siglen,
+                                       const unsigned char *tbs,
+                                       size_t tbslen))
+{
+    PKEY_DATA *kdata = NULL;
+    int rv = 0;
+
+    rv = pkey_test_init_keyctx(t, name, use_public);
+    if (t->skip || !rv)
+        return rv;
+    kdata = t->data;
+    kdata->keyopinit = keyopinit;
+    kdata->keyop = keyop;
+    kdata->init_controls = sk_OPENSSL_STRING_new_null();
+    kdata->controls = sk_OPENSSL_STRING_new_null();
+    return 1;
+}
+
+static int pkey_test_init_ex2(EVP_TEST *t, const char *name,
+                              int use_public,
+                              int (*keyopinit)(EVP_PKEY_CTX *ctx,
+                                               EVP_SIGNATURE *algo,
+                                               const OSSL_PARAM param[]),
+                              int (*keyop)(EVP_PKEY_CTX *ctx,
+                                           unsigned char *sig, size_t *siglen,
+                                           const unsigned char *tbs,
+                                           size_t tbslen))
+{
+    PKEY_DATA *kdata = NULL;
+    int rv = 0;
+    char algoname[OSSL_MAX_NAME_SIZE + 1];
+    char keyname[OSSL_MAX_NAME_SIZE + 1];
+    const char *p;
+
+    if ((p = strchr(name, ':')) == NULL
+        || p == name || p[1] == '\0' || p - name > OSSL_MAX_NAME_SIZE) {
+        TEST_info("Can't extract algorithm or key name from '%s'", name);
+        return 0;
+    }
+    strncpy(algoname, name, p - name);
```

Comment:
A bit of an abuse of `strncpy`. I expect the last arg here would normally be the size of the `algoname` array. Really you are using it for `memcpy`

Reply:
The size argument to strncpy() should be the number of non-NUL bytes you want to copy, surely?  But yeah, memcpy() works just as well, and if that looks better, then sure...

Reply:
I have always considered the final argument to strncpy to be the size of the destination buffer not the number of bytes you want to copy.